### PR TITLE
mds: reject lookup ino requests for mds dirs

### PR DESCRIPTION
**Pull request by batrick**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ukernel** [comment]
I'd like to forward to proper mds https://github.com/ceph/ceph/pull/38042/commits/a0f5dcb5423c1cb05c6fde1f7270a8480035867f

**Comment by jtlayton** [comment]
> I'd like to forward to proper mds [a0f5dcb](https://github.com/ceph/ceph/commit/a0f5dcb5423c1cb05c6fde1f7270a8480035867f)

@ukernel: we have a case where a client apparently tried to look up inode number 0x100 (See: https://tracker.ceph.com/issues/49922)

My understanding was that all of the inodes between MDS_INO_MDSDIR_OFFSET and MDS_INO_SYSTEM_BASE were for MDS internal usage only.  Is there ever a legitimate reason for a client to look up these inodes? If not, then it seems best to just pretend they don't exist if the client asks for them.

File: src/mds/Server.cc [hunk 1]
```diff
@@ -3899,6 +3899,11 @@ void Server::handle_client_lookup_ino(MDRequestRef& mdr,
     return _lookup_snap_ino(mdr);
 
   inodeno_t ino = req->get_filepath().get_ino();
+  auto _ino = ino.val;
+  if (MDS_INO_IS_STRAY(_ino) || MDS_INO_IS_MDSDIR(_ino)) {
```
Comment (jtlayton) [inline src/mds/Server.cc:3903]: How about this instead?
```
/* Don't allow client to look up inodes that are for MDS internal usage */
if (_ino < MDS_INO_SYSTEM_BASE && _ino != MDS_INO_ROOT) {
``` 
...that should catch any inode numbers in the MDS "private" range. Of course, if there is a legit reason to allow lookups of some of these inodes, then we will need to carve some others out.

**Review by jtlayton — CHANGES_REQUESTED** [review]

File: src/mds/Server.cc (hunk 1 above)
Comment (batrick) [reply src/mds/Server.cc:3903]: That also looks good. Let's see first what @ukernel says about  https://github.com/ceph/ceph/pull/40389#issuecomment-806632075

Comment (jtlayton) [reply src/mds/Server.cc:3903]: @ukernel: ping? Do you have any thoughts on the above? I'd like to add similar checks to the kernel (along with some stack traces so we can see what codepaths lead to these calls).

**Comment by lxbsz** [comment]
Should we also fix the `MDCache::do_open_ino_peer()` and `MDCache::handle_open_ino_reply()` ? To make sure that it won't try to open the inodes those are belonged to another specified MDS only, which will cause the inifinite loop here ?

**Comment by batrick** [comment]
> Should we also fix the `MDCache::do_open_ino_peer()` and `MDCache::handle_open_ino_reply()` ? To make sure that it won't try to open the inodes those are belonged to another specified MDS only, which will cause the inifinite loop here ?

Yes, I think so but the side-effects of that change were not easy to discern. This is the "quick" fix.

**Comment by lxbsz** [comment]
> > Should we also fix the `MDCache::do_open_ino_peer()` and `MDCache::handle_open_ino_reply()` ? To make sure that it won't try to open the inodes those are belonged to another specified MDS only, which will cause the inifinite loop here ?
> 
> Yes, I think so but the side-effects of that change were not easy to discern. This is the "quick" fix.

Okay.

File: src/mds/Server.cc (hunk 1 above)
Comment (lxbsz) [reply src/mds/Server.cc:3903]: There has another issue [1], which is the mdsdir `0x101` had issued caps to 3 clients and it prevented expiring the mdsdir inode in `mds.0`. 

In this case they were using both kclient and userspace client at the same time, so we also need a similar checks in `client/Client.cc`.

[1] https://tracker.ceph.com/issues/50112

Comment (jtlayton) [reply src/mds/Server.cc:3903]: I think this looks good, but I'd prefer we use the condition I suggested above. That's what we're doing in the client, so we probably ought to do the same in the MDS.

Comment (batrick) [reply src/mds/Server.cc:3903]: Done.

File: src/mds/Server.cc [hunk 2]
```diff
@@ -3899,6 +3899,11 @@ void Server::handle_client_lookup_ino(MDRequestRef& mdr,
     return _lookup_snap_ino(mdr);
 
   inodeno_t ino = req->get_filepath().get_ino();
+  auto _ino = ino.val;
+  if (_ino < MDS_INO_SYSTEM_BASE && _ino != MDS_INO_ROOT) {
+    respond_to_request(mdr, -CEPHFS_ESTALE);
+    return;
+  }
```
Comment (jtlayton) [inline src/mds/Server.cc:3906]: nit: a comment about why we do this wouldn't be a bad idea here, but looks good otherwise.

**Review by jtlayton — APPROVED** [review]

**Review by lxbsz — APPROVED** [review]

File: src/mds/Server.cc (hunk 2 above)
Comment (batrick) [reply src/mds/Server.cc:3913]: done

**Review by lxbsz — APPROVED** [review]

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-Apr-08

**Pull request merged**